File: inst/www/dragula-select.js
// ##############################
// ### dragulaSelectR options ###
// ##############################
var dragulaSelectR = {};

dragulaSelectR.options = {
  isContainer: function(el) {
    return el.classList.contains('ds-dragzone');
  },
  copy: function(el, source) {
    // Source -> Target only
    return source.classList.contains('ds-dragzone');
  },
  invalid: function (el, handle) {
    return ($(el).hasClass('ds-locked') || $(el).hasClass('ds-placeholder'));
  },
  accepts: function(el, target, source, sibling) {
    // Make sure option exists within dropzone
    let validOption = $(target).children(".ds-dropzone-options").children('.ds-dropoption[data-value="' + $(el).data('value') + '"]').length > 0;

    // Check if item is already there (not including temporary
    //   in-transit pre-drop item)
    let numItemsWithValue = $(target).children('[data-value="' + $(el).data('value') + '"]:not(".gu-transit")').length;
    let multivalued = $(target).hasClass('ds-multivalued');
    let spaceAvailable = ((multivalued || (numItemsWithValue === 0)) &&
                          (!$(target).hasClass('ds-max-input') ||
                           $(el).hasClass('ds-dropoption') ||
                           ($(target).hasClass('ds-replace-on-drop') && (sibling !== null))));

    // Source -> Target only AND
    //   no dropzone to different dropzone AND (note: caused issue when drop triggered before remove - might change in future)
    //   valid available option in dropzone AND
    //   not before a frozen item (note: need .gu-transit check as well)
    return (!target.classList.contains('ds-dragzone') &&
            !(source.classList.contains('ds-dropzone') && (source.id !== target.id)) &&
            validOption && spaceAvailable &&
            !$(sibling).is('.ds-freeze') && !$('.gu-transit').next().is('.ds-freeze'));
  },
  revertOnSpill: true, // Always revert to source container on spill
  removeOnSpill: true  // Always remove drag item on spill
};

function initDragulaSelectR() {
  dragulaSelectR.drake = dragula(dragulaSelectR.options);

  dragulaSelectR.drake.on("drag", function(el, source) {
    // Set removeOnSpill from source zone setting
    let removeOnSpill = $(source).data('remove-on-spill');
    dragulaSelectR.options.removeOnSpill = (removeOnSpill !== undefined ? removeOnSpill : true);
  });

  dragulaSelectR.drake.on("dragend", function(el) {
    // Set removeOnSpill back to default
    dragulaSelectR.options.removeOnSpill = true;
  });

  dragulaSelectR.drake.on("drop", function(el, target, source, sibling) {
    // Coming in from source - otherwise, do nothing
    if ($(el).hasClass('ds-dragitem')) {
      // Clone option with corresponding value
      let $newItem = $(target).children(".ds-dropzone-options").children('.ds-dropoption[data-value="' + $(el).data('value') + '"]').clone();

      // Update dropzone counter
      $(target).data('counter', $(target).data('counter') + 1);

      // Set instance id for new item (only used for multivalued)
      $newItem.attr('data-instance', $(target).hasClass('ds-multivalued') ? $(target).data('counter') : '');

      if (!$(target).hasClass('ds-hidden') && sibling) {
        $newItem.insertBefore(sibling);
      } else {
        $(target).append($newItem);
      }

      // Trigger selection if applicable
      if ($(target).data('select-on-drop')) {
        $newItem.trigger("click");
      }

      // Always remove element coming from source
      el.remove();

      // If replace on drop, remove replaced element (will always be next)
      if ($(target).hasClass('ds-max-input') && $(target).hasClass('ds-replace-on-drop')) {
        if (!$(target).hasClass('ds-hidden')) {
          // sibling should always exist
          sibling.remove();
        } else {
          $(target).children('.ds-dropoption').last().prev().remove();
        }
      }

      // Check if at max allowable inputs
      let numItemsTotal = $(target).children('.ds-dropoption:not(".gu-transit")').length;
      if (numItemsTotal === Number($(target).data('maxInput'))) {
        $(target).addClass('ds-max-input');
      }
    }

    // Raise an event to signal that the value changed
    $(target).trigger("change");
  });

  dragulaSelectR.drake.on("over", function(el, container, source) {
    $(container).addClass('gu-dragover');

    // Set direction (timed out due to glitch when over first occurs)
    setTimeout(function() {
      let direction = $(container).data('direction');
      dragulaSelectR.options.direction = (direction !== undefined ? direction : "vertical");
    });
    // Change content of item in transit
    $(el).html($(container).children(".ds-dropzone-options").children('.ds-dropoption[data-value="' + $(el).data('value') + '"]').html());

    // Hide placeholder if dropzone not hidden
    if (!$(container).hasClass("ds-hidden")) {
      $(container).children(".ds-placeholder").addClass("hidden");
    }
  });

  dragulaSelectR.drake.on("out", function(el, container, source) {
    $(container).removeClass('gu-dragover');

    // New entry. Dragula doesn't remove temporary gu-hide class for
    //   some reason.
    if (!$(source).data('remove-on-spill')) {
      $(el).removeClass("gu-hide");
    }

    // Un-hide placeholder if no more items in non-hidden dropzone
    if (!$(container).hasClass("ds-hidden") && ($(container).children('.ds-dropoption:not(".gu-transit")').length === 0)) {
      $(container).children(".ds-placeholder").removeClass("hidden");
    }
  });

  // Trigger changes on item removal
  dragulaSelectR.drake.on("remove", function(el, container, source) {
    $(container).removeClass('ds-max-input');

    if ($(source).hasClass('ds-dropzone')) {
      $(source).trigger("change");
    }

    let dzId = $(source).attr('id');
    if ($(el).hasClass('ds-selected')) {
      Shiny.onInputChange(dzId + "_selected", null);
    }
    if ($(el).hasClass('ds-invisible')) {
      Shiny.onInputChange(dzId + "_invisible", getValues($(source), '.ds-invisible'));
    }
    if ($(el).hasClass('ds-locked')) {
      Shiny.onInputChange(dzId + "_locked", getValues($(source), '.ds-locked'));
    }
  });
}

// #############################
// ### dragulaSelectR events ###
// #############################
$(document).on("ready", function() {
  if (dragulaSelectR.drake === undefined) {
    initDragulaSelectR();
  }
});

// #####################
// ### dropZoneInput ###
// #####################
var dropZoneBinding = new Shiny.InputBinding();

$.extend(dropZoneBinding, {
  find: function(scope) {
    return $(scope).find(".ds-dropzone");
  },
  initialize: function(el) {
    // Race condition
    if (dragulaSelectR.drake === undefined) {
      initDragulaSelectR();
    }
    dragulaSelectR.drake.containers.push(el);

    // Set multivalued counter to max instance value
    $(el).data('counter', Math.max(0, ...$('#' + el.id + ' > .ds-dropoption').map(function() { return this.dataset.instance })));

    // Selection is being made on a selectable zone
    if ($(el).hasClass('ds-selectable')) {
      $(el).on("click", ".ds-dropoption", function() {
        let $clicked = $(this);
        let dzId = $clicked.parent().attr('id');
        let newValue = optionValue($clicked.get(0));

        // Get currently selected - right now only one allowed
        let $selected = $clicked.siblings(".ds-selected");
        let currValue = ($selected.length ? optionValue($selected.get(0)) : null);

        $selected.removeClass("ds-selected");
        $clicked.addClass("ds-selected");
        Shiny.onInputChange(dzId + "_selected", newValue);
      });
    }

    // Toggle visibility
    $(el).on("click", ".ds-dropoption > .ds-toggle-visible i", function(ev) {
      ev.stopPropagation(); // Avoid selecting
      $(this).toggleClass("fa-eye fa-eye-slash");
      $(this).closest(".ds-dropoption").toggleClass("ds-invisible");
      let $dropzone = $(this).closest(".ds-dropzone");
      Shiny.onInputChange($dropzone.attr('id') + "_invisible", getValues($dropzone, '.ds-invisible'));
    });

    // Toggle draggability
    $(el).on("click", ".ds-dropoption > .ds-toggle-lock i", function(ev) {
      ev.stopPropagation(); // Avoid selecting
      $(this).toggleClass("fa-lock fa-lock-open");
      $(this).closest(".ds-dropoption").toggleClass("ds-locked");
      let $dropzone = $(this).closest(".ds-dropzone");
      Shiny.onInputChange($dropzone.attr('id') + "_locked", getValues($dropzone, '.ds-locked'));
    });

    // Initialize secondary inputs (selected, invisible, and locked).
    //   Need timeout so other binding stuff can happen before we update
    //   secondary inputs (essentially, without the timeout, no dice...)
    setTimeout(function() {
      let dzId = el.id;
      Shiny.onInputChange(dzId + "_selected", getValues($(el), '.ds-selected'));
      Shiny.onInputChange(dzId + "_invisible", getValues($(el), '.ds-invisible'));
      Shiny.onInputChange(dzId + "_locked", getValues($(el), '.ds-locked'));

      // Store settings to allow for server-side checking in updateDropZoneInput
      Shiny.onInputChange(dzId + "_settings:ds-fix-settings", {
        choices: $(el).find('.ds-dropzone-options > .ds-dropoption').map(function() { return optionValue(this) }).get(),
        multivalued: $(el).hasClass('ds-multivalued'),
        maxInput: $(el).data('max-input')
      });
    });
  },
  getValue: function(el) {
    return $('#' + el.id + ' > .ds-dropoption').map(function() { return optionValue(this) }).get();
  },
  subscribe: function(el, callback) {
    $(el).on("change.dropZoneBinding", function(e) {
      callback();
    });
  },
  unsubscribe: function(el) {
    $(el).off(".dropZoneBinding");
  },
  receiveMessage: function(el, data) {
    if (data.hasOwnProperty('action')) {
      if (data.action === "entangle") {
        // Replace drop options
        $('#' + el.id).children('.ds-dropoption').remove();
        $('#' + data.sourceId).children('.ds-dropoption').clone().appendTo('#' + el.id);

        // Copy counter information
        $('#' + el.id).data('counter', $('#' + data.sourceId).data('counter'));

        $(el).trigger("change");
      }
    }

    if (data.hasOwnProperty('placeholder')) {
      $(el).children(".ds-placeholder").html(data.placeholder);

      // Un-hide placeholder if no items in non-hidden dropzone
      if (!$(el).hasClass("ds-hidden") && ($(el).children('.ds-dropoption:not(".gu-transit")').length === 0)) {
        $(el).children(".ds-placeholder").removeClass("hidden");
      }
    }

    if (data.hasOwnProperty('presets')) {
      // Remove drop options
      $('#' + el.id).children('.ds-dropoption').remove();

      // Add new drop options
      Object.values(data.presets.values).forEach(function(value) {
        $(el).data('counter', $(el).data('counter') + 1);
        $(el).children(".ds-dropzone-options").children('.ds-dropoption[data-value="' + value + '"]').clone().attr("data-instance", $(el).hasClass('ds-multivalued') ? $(el).data('counter') : null).appendTo(el);
      });

      // Toggle placeholder status
      let numItemsTotal = Object.values(data.presets.values).length;
      if (numItemsTotal === 0) {
        $(el).children(".ds-placeholder").removeClass("hidden");
      } else {
        $(el).children(".ds-placeholder").addClass("hidden");
      }

      // Add proper class if at maximum input
      if (numItemsTotal === Number($(el).data('maxInput'))) {
        $(el).addClass('ds-max-input');
      } else {
        $(el).removeClass('ds-max-input');
      }

      $(el).trigger("change");
    }
  }
});

Shiny.inputBindings.register(dropZoneBinding);

// #########################
// ### Helpful functions ###
// #########################
function optionValue(el) {
  return [el.dataset.value, el.dataset.instance].filter(Boolean).join('-ds-');
}

function getValues($dropzone, withClass) {
  return $dropzone.children(withClass).map(function() { return optionValue(this) }).get();
}

File: R/dragulaSelect.R
#' dragulaSelectR: Utilizes Dragula JS library to implement a drag-and-drop Shiny select input.
#'
#' Utilizes Dragula JS library to implement a drag-and-drop Shiny select input.
#' Dragula is a drag-and-drop javascript library that is very lightweight and robust.
#' This implementation creates a Shiny input that replicates much of the functionality of
#' selectInput. Multiple zones for dragging and dropping are allowed. See the Dragula JS
#' library for more information (https://github.com/bevacqua/dragula).
#'
#' @examples
#' ## Only run examples in interactive R sessions
#' if (interactive()) {
#'
#' # basic example
#' shinyApp(
#'   ui = fluidPage(
#'     column(6,
#'       dragZone("dragzone", choices = list(one = "One",
#'                                           two = "Two",
#'                                           three = "Three",
#'                                           four = "Four"))
#'       ),
#'     column(6,
#'       dropZoneInput("dropzone", choices = list(one = "1",
#'                                                two = "2",
#'                                                three = "3",
#'                                                four = "4"),
#'                                 highlight = TRUE)
#'       )
#'   ),
#'   server = function(input, output) {
#'   }
#' )
#' }
#'
#' @docType package
#'
#' @import shiny
#'
#' @name dragulaSelectR
#'
#' @seealso \code{\link{dragZone}}, \code{\link{dropZoneInput}}
#'
NULL

#' Create a Dragula dragzone container
#'
#' @param id The container id.
#' @param choices List of values to select from.
#'
#' @export
#'
#' @examples
#' dragZone("dragzone", choices = list(one = "One",
#'                                     two = "Two",
#'                                     three = "Three",
#'                                     four = "Four"))
#'
#' @seealso \code{\link{dragulaSelectR}}
#'
dragZone <- function(id, choices) {

  if (missing(id)) {
    warning("Using generic 'dragzone' as id. HTML element may not be unique!")
    id <- "dragzone"
  }

  if (missing(choices)) {
    stop("You must specify choices for this dragZone. ")
  }

  # Resolve names
  choices <- choicesWithNames(choices)

  inputTag <- div(
    id = id,
    class = 'ds-dragzone',
    dragulaZoneItems('drag', 'options', choices)
  )

  attachDependencies(inputTag)
}

#' Create a Dragula dropzone input
#'
#' @param inputId The \code{input} slot that will be used to acces the value.
#' @param choices List of acceptable values with their associated labels. Note that
#'   the labels can be arbitrary HTML, as long as they are wrapped in a \code{tagList}.
#' @param presets Array or list of preset values.
#' @param hidden Should the selected items be hidden? This is useful to represent
#'   a reactive or event trigger.
#' @param placeholder If hidden is true, insert placeholder text.
#' @param highlight Highlights the container on dragover. Useful when \code{hidden} is active.
#' @param multivalued Allow multiple items with the same value?
#' @param selectable Are the items in this dropzone selectable? Default is \code{false}. Use
#'   Shiny input \code{input$<inputId>_selected} to access selected items.
#' @param selectOnDrop Should new dropped items be automatically selected?
#' @param togglevis Add an icon to allow toggling items between visible/invisible. Default is
#'   \code{false}. Use Shiny input \code{input$<inputId>_invisible} to access invisible items.
#' @param togglelock Add an icon to allow toggling items between locked/unlocked. Locked items
#'   are not draggable. Default is \code{false}. Use Shiny input \code{input$<inputId>_locked}
#'   to access locked items.
#' @param removeOnSpill Remove items when dragged outside dropzone? Default is \code{true}.
#' @param direction Direction (\code{horizontal} or \code{vertical}) to consider when
#'   determining where an element would be dropped. Default is \code{vertical}.
#' @param maxInput Maximum allowable dropped items.
#' @param replaceOnDrop Replace item on drop when at maximum allowable items?
#'
#' @export
#'
#' @examples
#' dropZoneInput("dropzone", choices = list(one = "1",
#'                                          two = "2",
#'                                          three = "3",
#'                                          four = "4"))
#'
#' @seealso \code{\link{dragulaSelectR}}
#'
dropZoneInput <- function(inputId, choices, presets=NULL, hidden=FALSE, placeholder=NULL,
                          highlight=FALSE, multivalued=FALSE, selectable=FALSE,
                          selectOnDrop=FALSE, togglevis=FALSE, togglelock=FALSE,
                          removeOnSpill=TRUE, direction="vertical", maxInput=Inf,
                          replaceOnDrop=FALSE) {

  # Resolve names
  choices <- choicesWithNames(choices)

  # Manage presets
  presets <- presetsWithOptions(presets, choices, multivalued)

  # Make sure number of preset values obeys maxInput setting
  if (length(presets$values) > maxInput) {
    stop("Number of preset values (", length(presets$values), ") exceeds the maximum allowable (", maxInput,")")
  }

  inputTag <- div(
    id = inputId,
    class = trimws(paste('ds-dropzone', opts2class(list(hidden = hidden,
                                                        highlight = highlight,
                                                        multivalued = multivalued,
                                                        selectable = selectable,
                                                        `max-input` = (length(presets$values) == maxInput),
                                                        `replace-on-drop` = replaceOnDrop))), "right"),
    "data-select-on-drop" = tolower(selectOnDrop),
    "data-remove-on-spill" = tolower(removeOnSpill),
    "data-direction" = tolower(direction),
    "data-max-input" = ifelse(is.infinite(maxInput), "Infinity", maxInput),
    insertPlaceholder(inputId, ifelse(hidden, placeholder, NA)),
    div(
      class = 'ds-dropzone-options',
      dragulaZoneItems('drop', 'options', choices,
                       togglevis = togglevis,
                       togglelock = togglelock)
    ),
    dragulaZoneItems('drop', 'presets',
                     items = presets$values,
                     ids = presets$ids,
                     selected = presets$selected,
                     invisible = presets$invisible,
                     locked = presets$locked,
                     freeze = presets$freeze,
                     togglevis = togglevis,
                     togglelock = togglelock)
  )

  attachDependencies(inputTag)
}

#' Create items for dragzones/dropzones
#'
#' In Dragula, the drag and drop areas are called "containers". This function
#' creates the individual draggable items and options in these containers.
#'
#' @param zone  Container zone type: either \code{drop} or \code{drag}
#' @param type  Are these \code{options} or \code{presets}?
#' @param items List of item labels, with names corresponding to values
#' @param ids If multivalued, these will be unique ids
#' @param selected  Selected items (array length of items - either NA or ds-selected)
#' @param invisible Invisible items (array length of items - either NA or ds-invisible)
#' @param locked  Locked items (array length of items - either NA or ds-locked)
#' @param freeze No items allowed before these. Analogous to freezing the first few
#'   columns of a spreadsheet (array length of items - either NA or ds-freeze). Makes
#'   since only for first initial items, and when used in conjunction with locked.
#'   Frozen items also cannot be toggled.
#' @param togglevis Add an icon to allow toggling items between visible/invisible.
#' @param togglelock Add an icon to allow toggling items between locked/unlocked.
#'
#' @return div element
#'
dragulaZoneItems <- function(zone, type, items, ids=rep(NA, length(items)), selected=NULL,
                             invisible=NULL, locked=NULL, freeze=NULL, togglevis=FALSE, togglelock=FALSE) {
  if (!(zone %in% c('drag', 'drop'))) {
    stop(zone, " is not a valid container type. Dragula container type must be either 'drag' or 'drop'")
  }
  if (!(type %in% c('options', 'presets'))) {
    stop(type, " is not a valid item type. Item type must be either 'options' or 'presets'")
  }
  values <- names(items)
  tagList(
    lapply(seq_along(items),
           FUN = function(values, labels, ids, selected, invisible, locked, freeze, togglevis, togglelock, i) {
             div(
               "data-value" = values[[i]] %||% labels[[i]],
               "data-instance" = ids[[i]],
               class = trimws(paste(paste0('ds-', ifelse(zone=='drop', 'dropoption', 'dragitem')),
                                    paste(keepTruthy(c(selected[[i]], invisible[[i]], locked[[i]], freeze[[i]])), collapse = ' '))
                              ),
               labels[[i]] %||% values[[i]],
               switch(togglevis && (zone == 'drop') && !isTruthy(freeze[[i]]),
                      div(class = "ds-toggle-visible",
                          ifelse(isTruthy(invisible[[i]]), tagList(icon("eye-slash")), tagList(icon("eye")))),
                      NULL),
               switch(togglelock && (zone == 'drop') && !isTruthy(freeze[[i]]),
                      div(class = "ds-toggle-lock",
                          ifelse(isTruthy(locked[[i]]), tagList(icon("lock")), tagList(icon("lock-open")))),
                      NULL)
             )
           }, values = values, labels = items, ids = ids, selected = selected, invisible = invisible,
           locked = locked, freeze = freeze, togglevis = togglevis, togglelock = togglelock
    )
  )
}

#' Run dragulaSelectR Example Applications
#'
#' Launch dragulaSelectR example applications, and optionally, your system's web browser.
#'
#' @param example The name of the example to run, or \code{NA} (the default) to
#'   list the available examples.
#' @param port The TCP port that the application should listen on. Defaults to
#'   choosing a random port.
#' @param launch.browser If true, the system's default web browser will be
#'   launched automatically after the app is started. Defaults to true in
#'   interactive sessions only.
#' @param host The IPv4 address that the application should listen on. Defaults
#'   to the \code{shiny.host} option, if set, or \code{"127.0.0.1"} if not.
#' @param display.mode The mode in which to display the example. Defaults to
#'   \code{showcase}, but may be set to \code{normal} to see the example without
#'   code or commentary.
#'
#' @examples
#' ## Only run this example in interactive R sessions
#' if (interactive()) {
#'   # List all available examples
#'   runExample()
#'
#'   # Run one of the examples
#'   runExample("01_basic")
#'
#'   # Print the directory containing the code for all examples
#'   system.file("examples", package="dragulaSelectR")
#' }
#' @export
#'
#' @references \code{\link[shiny]{runExample}}
#'
runExample <- function(example=NA,
                       port=NULL,
                       launch.browser=getOption('shiny.launch.browser',
                                                interactive()),
                       host=getOption('shiny.host', '127.0.0.1'),
                       display.mode=c("auto", "normal", "showcase")) {
  examplesDir <- system.file('examples', package='dragulaSelectR')
  dir <- resolve(examplesDir, example)
  if (is.null(dir)) {
    if (is.na(example)) {
      errFun <- message
      errMsg <- ''
    }
    else {
      errFun <- stop
      errMsg <- paste('Example', example, 'does not exist. ')
    }

    errFun(errMsg,
           'Valid examples are "',
           paste(list.files(examplesDir), collapse='", "'),
           '"')
  }
  else {
    runApp(dir, port = port, host = host, launch.browser = launch.browser,
           display.mode = display.mode)
  }
}

# HELPERS ----

#' Attach javascript dependencies
#'
#' \code{attachDependencies} attaches the javascript dependencies.Specifically,
#' the Dragula JS package is attached, as well as the javascript wrapper and
#' input bindings.
#'
#' @param ... Shiny tag object
attachDependencies <- function(...) {
  deps <- list(
    htmltools::htmlDependency(name = "dragula", version = "3.7.2",
                              package = "dragulaSelectR",
                              src = "www/dragula-3.7.2",
                              script = "dragula.min.js",
                              stylesheet = "dragula.min.css"
    ),
    htmltools::htmlDependency(name = "dragula-select", version = "0.0.0.9000",
                              package = "dragulaSelectR",
                              src = "www",
                              script = "dragula-select.js",
                              stylesheet = "dragula-select.css"
    )
  )
  htmltools::attachDependencies(..., deps)
}

#' Converts options to class names
#'
#' @param varArgs Named list of options for dropzone
#'
#' @return String
#'
opts2class <- function(varArgs) {
  varArgsNames <- names(varArgs)[vapply(varArgs, isTRUE, logical(1))]
  paste(lapply(varArgsNames, function(opt) { paste0('ds-', opt) }), collapse = ' ')
}

#' Input is set of possible unique ids from multivalued inputs
#'
#' @param values Values returned by dropzone input
#'
#' @return logical
#'
#' @export
isMultivalued <- function(values) {
  nvals <- length(values)
  if (is.null(values))
    return(FALSE)

  # Not unique values
  if (length(unique(values)) != nvals)
    return(FALSE)

  # Doesn't have the -ds- separator in all entries
  if (length(grep('-ds-', values)) != nvals)
    return(FALSE)

  # Check individual id entries - are they integers when casted?
  ids <- multivalues(values, ids=TRUE)
  if (anyNAOrFalse(isWholeNum(suppressWarnings(as.double(ids)))))
    return(FALSE)

  return(TRUE)
}

#' Convert unique values to multivalues
#'
#' (Multivalued dropzones only) This will drop the added unique counter ids
#' and convert to multivalued inputs.
#'
#' @param values Values returned by dropzone input
#' @param ids Return unique ids rather than multivalues?
#'
#' @return Multivalues with unique id stripped away.
#'
#' @export
multivalues <- function(values, ids=FALSE) {
  if (!is.null(values)) {
    sapply(strsplit(values, '-ds-'),
           FUN = function(x) {
             if (length(x) == 1) {
               return(ifelse(ids, NA, x))
             } else {
               return(ifelse(ids, x[length(x)], paste(x[-length(x)], collapse = '-ds-')))
             }
           }, simplify = "array")
  }
}

#' Entangle two dropzones
#'
#' Create observe events that entangle two Shiny dropzones. Useful for
#' hidden dropzones that take drops but display options elsewhere.
#'
#' @param session The \code{session} object passed to function given to \code{shinyServer}.
#' @param dropZoneOneId  The \code{id} of the first dropzone.
#' @param dropZoneTwoId  The \code{id} of the second dropzone.
#'
#' @return Expression including the two observe events.
#' @export
entangle <- function(session, dropZoneOneId, dropZoneTwoId) {
  return({
    observeEvent(session$input[[dropZoneOneId]], {
      if (!isTRUE(all.equal(session$input[[dropZoneOneId]], session$input[[dropZoneTwoId]]))) {
        entangleSourceToTarget(session, sourceId = dropZoneOneId, targetId = dropZoneTwoId)
      }
    }, ignoreNULL = FALSE)

    observeEvent(session$input[[dropZoneTwoId]], {
      if (!isTRUE(all.equal(session$input[[dropZoneOneId]], session$input[[dropZoneTwoId]]))) {
        entangleSourceToTarget(session, sourceId = dropZoneTwoId, targetId = dropZoneOneId)
      }
    }, ignoreNULL = FALSE)
  })
}

#' Update target dropzone entangled with source dropzone
#'
#' @param session The \code{session} object passed to function given to \code{shinyServer}.
#' @param sourceId  The \code{id} of the source dropzone.
#' @param targetId  The \code{id} of the target dropzone.
#'
entangleSourceToTarget <- function(session, sourceId, targetId) {
  message <- dropNulls(list(action = "entangle", sourceId = session$ns(sourceId)))
  session$sendInputMessage(targetId, message)
}
